org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/hover/JvmAnnotationReferencePrinter.java

```diff
@@ -0,0 +1,220 @@
+			buffer.append(")");
+		}
+
+		return buffer.toString();
+	}
+
+	protected String _internalToString(JvmAnnotationValue jvmAnnotationValue) {
+		EStructuralFeature ref = jvmAnnotationValue.eClass().getEStructuralFeature("values");
+		if (ref == null) {
+			throw new IllegalStateException("Cannot find feature with name 'values' in " + jvmAnnotationValue);
+		}
+
+		EList<?> values = (EList<?>) jvmAnnotationValue.eGet(ref);
+		if (values.size() > 1) {
+			return "#[" + values.stream().map((value) -> internalToString(value)).collect(Collectors.joining(", ")) + "]";
+		} else {
+			return internalToString(values.get(0));
+		}
+	}
+
+	protected String _internalToString(Object obj) {
+		return obj.toString();
+	}
+
+	protected String _internalToString(XTypeLiteral typeLiteral) {
+		URI uri = EcoreUtil.getURI(typeLiteral.getType());
+		String text = createLinkWithLabel(XtextElementLinks.XTEXTDOC_SCHEME, uri, typeLiteral.getType().getSimpleName());
+		return text + String.join("", typeLiteral.getArrayDimensions());
+	}
+
+	protected String _internalToString(XListLiteral listLiteral) {
+		return "#[" + listLiteral.getElements().stream().map(element -> internalToString(element)).collect(Collectors.joining(", ")) + "]";
+	}
+
+	protected String _internalToString(XBinaryOperation binaryOperation) {
+		return internalToString(binaryOperation.getLeftOperand()) + " " + binaryOperation.getConcreteSyntaxFeatureName() + " "
+				+ internalToString(binaryOperation.getRightOperand());
+	}
+
+	protected String _internalToString(XStringLiteral stringLiteral) {
+		return "\"" + stringLiteral.getValue() + "\"";
+	}
+
+	protected String _internalToString(XNumberLiteral numberLiteral) {
+		return numberLiteral.getValue();
+	}
+
+	protected String _internalToString(XBooleanLiteral booleanLiteral) {
+		return Boolean.valueOf(booleanLiteral.isIsTrue()).toString();
+	}
+
+	protected String _internalToString(XMemberFeatureCall memberFeatureCall) {
+		if (memberFeatureCall.isPackageFragment()) {
+			return null;
+		}
+		String prefix = internalToString(memberFeatureCall.getMemberCallTarget());
+		return internalHandleAbstractFeatureCall(prefix, memberFeatureCall);
+	}
+
+	protected String _internalToString(XFeatureCall featureCall) {
+		return featureCall.isPackageFragment() ? null : internalHandleAbstractFeatureCall(null, featureCall);
+	}
+
+	protected String internalHandleAbstractFeatureCall(String prefix, XAbstractFeatureCall abstractFeatureCall) {
+		String postfix = null;
+		if (abstractFeatureCall.getFeature() != null && !abstractFeatureCall.getFeature().eIsProxy()) {
+			URI uri = EcoreUtil.getURI(abstractFeatureCall.getFeature());
+			postfix = createLinkWithLabel(XtextElementLinks.XTEXTDOC_SCHEME, uri, abstractFeatureCall.getConcreteSyntaxFeatureName());
+		} else {
+			postfix = abstractFeatureCall.getConcreteSyntaxFeatureName();
+		}
+
+		if (prefix == null) {
+			return postfix;
+		} else {
+			return prefix + "." + postfix;
```

Comment:
This will look a little weird if the postfix is null since the link is broken.

Reply:
Would you recommend logging and error message if the `postfix` is null, before returning the value?

Reply:
No, I don't think an error is necessary. If there is an EMF proxy involved somewhere, this would be revealed to the user as a linking diagnostic or by means of a validation.

Reply:
I understand. Do you want me to keep the code as it is?

Reply:
Ok